build/build.sh

```diff
@@ -0,0 +1,374 @@
+#!/bin/bash
+
+build=false
+ci=false
+configuration="Debug"
+help=false
+nolog=false
+pack=false
+prepareMachine=false
+rebuild=false
+norestore=false
+sign=false
+skipTests=false
+bootstrapOnly=false
+verbosity="minimal"
+hostType="core"
+properties=""
+
+function Help() { 
+  echo "Common settings:"
+  echo "  -configuration <value>  Build configuration Debug, Release"
+  echo "  -verbosity <value>      Msbuild verbosity (q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic])"
+  echo "  -help                   Print help and exit"
+  echo ""
+  echo "Actions:"
+  echo "  -norestore              Don't automatically run restore"
+  echo "  -build                  Build solution"
+  echo "  -rebuild                Rebuild solution"
+  echo "  -skipTests              Don't run tests"
+  echo "  -bootstrapOnly          Don't run build again with bootstrapped MSBuild"
+  echo "  -sign                   Sign build outputs"
+  echo "  -pack                   Package build outputs into NuGet packages and Willow components"
+  echo ""
+  echo "Advanced settings:"
+  echo "  -ci                     Set when running on CI server"
+  echo "  -nolog                  Disable logging"
+  echo "  -prepareMachine         Prepare machine for CI run"
+  echo ""
+  echo "Command line arguments not listed above are passed through to MSBuild."
+}
+
+while [[ $# > 0 ]]; do
+  lowerI="$(echo $1 | awk '{print tolower($0)}')"
+  case $lowerI in
+    -build)
+      build=true
+      shift 1
+      ;;
+    -ci)
+      ci=true
+      shift 1
+      ;;
+    -configuration)
+      configuration=$2
+      shift 2
+      ;;
+    -help)
+      Help
+      exit 0
+      ;;
+    -nolog)
+      nolog=true
+      shift 1
+      ;;
+    -pack)
+      pack=true
+      shift 1
+      ;;
+    -preparemachine)
+      prepareMachine=true
+      shift 1
+      ;;
+    -rebuild)
+      rebuild=true
+      shift 1
+      ;;
+    -norestore)
+      norestore=true
+      shift 1
+      ;;
+    -sign)
+      sign=true
+      shift 1
+      ;;
+    -skiptests)
+      skipTests=true
+      shift 1
+      ;;
+    -bootstraponly)
+      bootstrapOnly=true
+      shift 1
+      ;;
+    -verbosity)
+      verbosity=$2
+      shift 2
+      ;;
+    *)
+      properties="$properties $1"
+      shift 1
+      ;;
+  esac
+done
+
+function CreateDirectory {
+  if [ ! -d "$1" ]
+  then
+    mkdir -p "$1"
+  fi
+}
+
+function GetLogCmd {
+  if $ci || $log
+  then
+    CreateDirectory $LogDir
+
+    local logCmd="/bl:$LogDir/$1.binlog"
+
+    # When running under CI, also create a text log, so it can be viewed in the Jenkins UI
+    if $ci
+    then
+      logCmd="$logCmd /fl /flp:Verbosity=diag;LogFile=\"$LogDir/$1.log\""
+    fi
+  else
+    logCmd=""
+  fi
+
+  echo $logCmd
+}
+
+function CallMSBuild {
+  local commandLine=""
+
+  if [ -z $msbuildHost ]
+  then
+    commandLine="$msbuildToUse $*"
+  else
+    commandLine="$msbuildHost $msbuildToUse $*"
```

Comment:
If args CallMSBuild includes paths with spaces, this line will build incorrect command.